fix weight norm backward bug on CPU when OMP_NUM_THREADS <= 2

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80930

fix https://github.com/pytorch/pytorch/issues/80569
root cause: `weight_norm_backward_last_dim_kernel` will create a temp buffer to
do vertical reduction, size of [num_threads, N] (N is the size of last dimension of v)

to save additional memory allocation, the original kernel reuse the buffer after
the vertical sum:
  1st row stores the final result of sum
  2nd row stores coefficient a
  3rd row stores coefficient b

when OMP_NUM_THREADS <=2, this will cause illegal memory access since the buffer size
will be only 1*N or 2*N;

the fix is use a separate buffer (`a_b`) to store the coefficients of a and b.

mingfeima committed:
fix weight norm backward bug on CPU when OMP_NUM_THREADS <= 2

fix https://github.com/pytorch/pytorch/issues/80569
root cause: `weight_norm_backward_last_dim_kernel` will create a temp buffer to
do vertical reduction, size of [num_threads, N] (N is the size of last dimension of v)

to save additional memory allocation, the original kernel reuse the buffer after
the vertical sum:
  1st row stores the final result of sum
  2nd row stores coefficient a
  3rd row stores coefficient b

when OMP_NUM_THREADS <=2, this will cause illegal memory access since the buffer size
will be only 1*N or 2*N;

the fix is use a separate buffer (`a_b`) to store the coefficients of a and b.

[ghstack-poisoned]

mingfeima commented:
This PR is to fix https://github.com/pytorch/pytorch/issues/80569.

To reproduce the issue, need to set `OMP_NUM_THREADS`=1 or 2 (3 and above will pass)

I have reproduced the issue on our local machine:
```
### before:
(pytorch-mingfei) [mingfeim@mlt-skx089 op_bench-py]$ OMP_NUM_THREADS=2 python test_weight_norm.py
*** Error in `python': free(): invalid next size (fast): 0x000055fc274ef6c0 ***
======= Backtrace: =========
/usr/lib64/libc.so.6(+0x7c619)[0x7f9d6c447619]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch_python.so(+0x3f27fe)[0x7f9d5d6db7fe]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libc10.so(_ZN3c1010TensorImplD0Ev+0x9)[0x7f9d53c473e9]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so(+0x3f5acdb)[0x7f9d57dfbcdb]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so(_ZN2at6native24weight_norm_backward_cpuERKNS_6TensorES3_S3_S3_l+0x130)[0x7f9d5541fb00]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so(+0x1d4185c)[0x7f9d55be285c]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so(+0x342eeb4)[0x7f9d572cfeb4]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so(+0x342e403)[0x7f9d572cf403]
/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so(_ZN2at4_ops31_weight_norm_interface_backward4callERKNS_6TensorES4_S4_S4_l+0x38a)[0x7f9d55a18eda]

....

### after:
(pytorch-mingfei) [mingfeim@mlt-skx089 op_bench-py]$ OMP_NUM_THREADS=2 python test_weight_norm.py
tensor(0.9946, grad_fn=<MseLossBackward0>)
```

frank-wei commented:
Thanks @mingfeima  for the fixing. The solution looks good to me.
cc @malfet  @albanD

malfet commented on aten/src/ATen/native/cpu/WeightNormKernel.cpp:
```diff
@@ -332,6 +332,11 @@ void weight_norm_backward_last_dim_kernel(
   auto buffer_data = buffer.data_ptr<accscalar_t>();
 
+  // 1st row to store coefficient a and 2nd row to store coefficient b
+  Tensor a_b = at::empty({2, N}, saved_norm.options()).zero_();
```
Hmm, wouldn't be more convenient to just set 1st dimension of buffer tensor to `std::max(2, N)`

mingfeima replied:
Good idea! Updated!

malfet reviewed (approved):
Looks good to me, but please consider getting rid of additional buffer and just tweak dimensions of the first one.

mingfeima committed:
Update on "fix weight norm backward bug on CPU when OMP_NUM_THREADS <= 2"


fix https://github.com/pytorch/pytorch/issues/80569
root cause: `weight_norm_backward_last_dim_kernel` will create a temp buffer to
do vertical reduction, size of [num_threads, N] (N is the size of last dimension of v)

to save additional memory allocation, the original kernel reuse the buffer after
the vertical sum:
  1st row stores the final result of sum
  2nd row stores coefficient a
  3rd row stores coefficient b

when OMP_NUM_THREADS <=2, this will cause illegal memory access since the buffer size
will be only 1*N or 2*N;

the fix is use a separate buffer (`a_b`) to store the coefficients of a and b.

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f
